### PR TITLE
PLT-5750 Add sequence number to websocket connections and events

### PR DESCRIPTION
**Pull request by jwilander**
#### Summary
* Remove the pre computation of websocket event JSON
* Changed WebSocketEvents to be passed as value (to prevent race condition when adding sequence number)
* Add a sequence number to each websocket connection
* Client will run reconnect logic if it gets an out of order sequence

I'm aware some of this may have a significant effect on performance. @crspeller is going to run load tests but in the interest of trying this out in a real environment we're going to merge it now and revert it if we need to.

#### Ticket Link
https://mattermost.atlassian.net/browse/PLT-5750

### CODE REVIEW

File: app/web_conn.go [hunk 1]
```diff
@@ -35,6 +35,7 @@ type WebConn struct {
 	Locale                    string
 	AllChannelMembers         map[string]string
 	LastAllChannelMembersTime int64
+	Sequence                  uint64
```
Comment (coreyhulen) [inline app/web_conn.go]: Minor nit why uint vs int?

**Comment by coreyhulen** [comment]
I think it should still be pass by pointer everywhere except when it sends the message to the write pump, then it can be copied.  I think the only place that would need to change would be https://github.com/mattermost/platform/blob/master/app/web_hub.go#L297 ?

Or copy the msg in the writePump then set the sequence id correctly then send it.  This would be even better since it's doing the copy on the writePump thread.

File: app/web_conn.go [hunk 2]
```diff
@@ -104,8 +105,17 @@ func (c *WebConn) WritePump() {
 				return
 			}
 
+			var msgBytes []byte
```
Comment (coreyhulen) [inline app/web_conn.go]: Copy clone/copy the point to model.WebSocketEvent right here then set the sequence number, then send it ?  That way it's only copied once.

File: app/web_conn.go (hunk 1 above)
Comment (jwilander) [inline app/web_conn.go]: No reason, it's always going to be positive so I went with uint64, I'm 0/5 though

File: app/web_conn.go (hunk 2 above)
Comment (jwilander) [inline app/web_conn.go]: Yeah I wasn't sure if that was any better than passing it down as value but I can do that instead

**Comment by jwilander** [comment]
@coreyhulen made the requested changes

**Comment by VeraLyu** [comment]
Hi @jwilander ,
I may get it wrong but as far as I know, websocket is based on TCP instead of HTTP and msg order is guaranteed by TCP connection. So could you let me know what usecase this patch covers? Thanks!

**Comment by jwilander** [comment]
This isn't to preserve message order, it is to detect when the client misses events when temporarily losing network connection (think walking out of range or switching networks). When we detect this happening the client knows it likely missed a new post and can fetch latest

**Comment by VeraLyu** [comment]
In this situation of sudden lost of connection, server TCP re-send package as it did not get ACK, and client reorders received packages in TCP window to its sending order meaning we won't miss anything. Or if the TCP connection permanently lost, we'll never receive package on this ws again. I'm still feel confused...

**Comment by VeraLyu** [comment]
I drew a pic to make my point clearer, here we have 4 user scenarios: A. stable connection B. unstable connection C. lost connection D. reconnected. I think we won't miss any events in A.B.D(I already explained why B won't lose msg in comment above), the only chance to miss event/msg is C. And as you already did in previous patches of syncing missing msgs in the beginning of phase D, I think we should already avoid other chances of missing events. That is why  I can't think out a usecase for sequence NO. Pls let me know what you think. Thanks a lot!
ws

**Comment by jwilander** [comment]
Thanks your interest in this @VeraLyu 

The crux of the matter is that we need a guarantee that the event message is getting from the application layer on the server to the application layer on the client. TCP handles that at the transport layer, but even there with a broken network connection it's possible to lose messages even with TCP attempting to resend packets. So this is our method at the application layer to detect if we missed message for whatever reason.

**Comment by jwilander** [comment]
@coreyhulen I added the extra logging and also specified a reconnect callback on the websocket client instead of only relying on the one in the client

**Pull request merged**